### PR TITLE
Add metadataSufficient status to indicate wether cluster have get all api-resource 

### LINKED ISSUE #1673: Join cluster failed（The logic to determine cluster health may not quite right  ）
**What happened**:
The ready status of cluster is false，and the output of `describe` is below
图片

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

1. step1: apply a apiservice in member
 
```yaml
# apiservice.yaml
apiVersion: apiregistration.k8s.io/v1
kind: APIService
metadata:
  name: v1alpha1.cluster.karmada.io
  labels:
    app: karmada-aggregated-apiserver
    apiserver: "true"
spec:
  insecureSkipTLSVerify: true
  group: cluster.karmada.io
  groupPriorityMinimum: 2000
  service:
    name: karmada-aggregated-apiserver
    namespace: karmada-system
  version: v1alpha1
  versionPriority: 10
```
2.step2: join the member cluster
```
kubectl karmada join member47 --kubeconfig=/root/.kube/config --karmada-context=karmada-apiserver --cluster-kubeconfig=/root/.kube/config --cluster-context=cluster47
``` 

**Anything else we need to know?**:

when I excute the command `kubectl api-resources`,the output is 
图片
But I don't think this should lead to an unhealthy state of the cluster. Other functions are normal, do not affect the use

Code location at `pkg/controllers/status/cluster_status_controller.go` of line 171
```
	apiEnables, err := getAPIEnablements(clusterClient)
	if err != nil {
		return c.setStatusCollectionFailedCondition(cluster, currentClusterStatus, fmt.Sprintf("failed to get the list of APIs installed in the member cluster: %v", err))
	}
```
Maybe we should ignore the err? I am not sure,asking you for response.

### PR DESCRIPTION
**Pull request by duanmengkk**
Signed-off-by: duanmeng <duanmeng_yewu@cmss.chinamobile.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #1673

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by duanmengkk** [comment]
/retitle Add warn status of the member cluster when failed get all supported resources from server

**Comment by duanmengkk** [comment]
/assign @XiShanYongYe-Chang
before 
https://github.com/karmada-io/karmada/issues/1673
now 
image
image

File: pkg/apis/cluster/v1alpha1/types.go [hunk 1]
```diff
@@ -127,6 +127,9 @@ type LocalSecretReference struct {
 const (
 	// ClusterConditionReady means the cluster is healthy and ready to accept workloads.
 	ClusterConditionReady = "Ready"
+	// ClusterConditionWarn means some err occurs when access member cluster but it doesn't affect the use of member cluster
+	// ClusterConditionWarn are ready to accept workloads too
+	ClusterConditionWarn = "Warn"
```
Comment (RainbowMango) [inline pkg/apis/cluster/v1alpha1/types.go:132]: `Warn` seems too common for me.
Let's try to work out a better one.

Comment (duanmengkk) [reply pkg/apis/cluster/v1alpha1/types.go:132]: What is your suggestion?How about `alarm`,I am limited by my english level.

Comment (duanmengkk) [reply pkg/apis/cluster/v1alpha1/types.go:132]: > `APIEnable`?

`warn `means control plane can not get all api list in member cluster now,perhaps a condition names `APIUnreachable` ?

Comment (RainbowMango) [reply pkg/apis/cluster/v1alpha1/types.go:132]: @Tingtal 
Need your help here.

The Karmada system will gather some information from a cluster, such as `version`, `apis`, etc. 
We want to define a `state` to describe if all required information is collected. 

Assume the `state` is `A`:
- A= true means all required information is collected.
- B= false means some of the required information is missed.

Comment (RainbowMango) [reply pkg/apis/cluster/v1alpha1/types.go:132]: Post an option(brainstorming):
`positive`: 
- `true` means we have enough confidence with the cluster based on sufficient information we got.
- `false` means lack of confidence as we cant know the cluster clearly as missing information.

Comment (RainbowMango) [reply pkg/apis/cluster/v1alpha1/types.go:132]: another option:
`MetadataSufficient`: 
- `true`: We get all the required data
- `false`: We didn't get all the required data, some of them are missing.

Comment (Tingtal) [reply pkg/apis/cluster/v1alpha1/types.go:132]: Ideas to share:

dataSufficient
dataIntact
requiredDataObtained
requiredDataReady
dataAllObtained
dataAllReady
inPlace

Comment (duanmengkk) [reply pkg/apis/cluster/v1alpha1/types.go:132]: `MetadataSufficient `+1

Comment (RainbowMango) [reply pkg/apis/cluster/v1alpha1/types.go:132]: How about talking about this and deciding at the next community meeting? @duanmengkk 

More work to do after the condition I guess.

Comment (duanmengkk) [reply pkg/apis/cluster/v1alpha1/types.go:132]: ok,let's discuss it

Comment (RainbowMango) [reply pkg/apis/cluster/v1alpha1/types.go:132]: Fine. thanks. please remember to add an agenda to the meeting notes.

Comment (duanmengkk) [reply pkg/apis/cluster/v1alpha1/types.go]: The latest code judge the health status of the cluster by `/readyz` and `/healthz` and ignore the error when `get pod` and `get node`,`get api-resource` ,just log the error and do not fail over.
So,should I close the PR ? @RainbowMango
image

**Comment by duanmengkk** [comment]
image
/cc @dddddai if cluster is oline and healty ,but `get pod` and `get node` is failed, should not we handle  the error and mark cluster status as unready?

**Comment by RainbowMango** [comment]
>  if cluster is oline and healty ,but get pod and get node is failed, should not we handle the error and mark cluster status as unready?

I don't think so.

**Comment by dddddai** [comment]
> > if cluster is oline and healty ,but get pod and get node is failed, should not we handle the error and mark cluster status as unready?
> 
> I don't think so.

+1, I think the ready condition should only represent the health check result

File: pkg/apis/cluster/v1alpha1/types.go [hunk 2]
```diff
@@ -127,6 +127,9 @@ type LocalSecretReference struct {
 const (
 	// ClusterConditionReady means the cluster is healthy and ready to accept workloads.
 	ClusterConditionReady = "Ready"
+	// ClusterMetadataSufficient means some err occurs when access member cluster but it doesn't affect the use of member cluster
+	// ClusterMetadataSufficient are ready to accept workloads too
+	ClusterMetadataSufficient = "MetadataSufficient"
```
Comment (RainbowMango) [inline pkg/apis/cluster/v1alpha1/types.go:132]:
```suggestion
	// ClusterMetadataSufficient means some metadata of the cluster is unavailable, such as Kubernetes version, or
	// API enablement list.
	// Clusters in insufficient mode condition(MetadataSufficient=false) are still ready to accept workloads,
	// but the scheduler might be favor to cluster in sufficient condition(MetadataSufficient=true), that depends on
	// scheduler implementation.
	ClusterMetadataSufficient = "MetadataSufficient"
```
How about this? 
(And, don't get me wrong, I don't mean this PR should take care of the scheduler, this PR just focus on how to maintain the condition.)

Comment (duanmengkk) [reply pkg/apis/cluster/v1alpha1/types.go]: fixed

File: pkg/controllers/status/cluster_status_controller.go [hunk 3]
```diff
@@ -47,6 +47,9 @@ const (
 	clusterNotReachableReason = "ClusterNotReachable"
 	clusterNotReachableMsg    = "cluster is not reachable"
 	statusCollectionFailed    = "StatusCollectionFailed"
+	metadataSufficientMsg     = "all APIs installed in the cluster have been obtained"
```
Comment (RainbowMango) [inline pkg/controllers/status/cluster_status_controller.go:50]: metadata sufficient not only include `APIs`, see the condition comments.

File: pkg/printers/internalversion/printers.go [hunk 4]
```diff
@@ -18,6 +18,7 @@ func AddHandlers(h printers.PrintHandler) {
 		{Name: "Version", Type: "string", Description: "KubernetesVersion represents version of the member cluster."},
 		{Name: "Mode", Type: "string", Description: "SyncMode describes how a cluster sync resources from karmada control plane."},
 		{Name: "Ready", Type: "string", Description: "The aggregate readiness state of this cluster for accepting workloads."},
+		{Name: "MetadataSufficient", Type: "string", Description: "The status indicate that all APIs are obtained"},
```
Comment (RainbowMango) [inline pkg/printers/internalversion/printers.go:21]: Is this used to format the output for `kubectl get` ?

Is there a test report?

File: pkg/util/membercluster_client.go [hunk 5]
```diff
@@ -31,6 +32,12 @@ type DynamicClusterClient struct {
 	ClusterName      string
 }
 
+// AggregatorClusterClient stands for a aggregator client for the given member cluster
+type AggregatorClusterClient struct {
+	AggregatorClient *aggregatorclient.Clientset
+	ClusterName      string
+}
```
Comment (RainbowMango) [inline pkg/util/membercluster_client.go:39]: I guess this is a discovery client, right? 
I'm not sure if we really need this, (`clusterClient.KubeClient.Discovery().ServerGroupsAndResources()` can't help? )might be @XiShanYongYe-Chang could take a look.

File: pkg/controllers/status/cluster_status_controller.go [hunk 6]
```diff
@@ -361,11 +377,16 @@ func getKubernetesVersion(clusterClient *util.ClusterClient) (string, error) {
...
+// getAPIEnablements get the list of APIs installed in the member cluster
 func getAPIEnablements(clusterClient *util.ClusterClient) ([]clusterv1alpha1.APIEnablement, error) {
+	// This can return an error *and* the results it was able to find.
+	// If the length of api resources is not zero,we don't need to fail on the error.
 	_, apiResourceList, err := clusterClient.KubeClient.Discovery().ServerGroupsAndResources()
-	if err != nil {
+	if len(apiResourceList) == 0 {
+		klog.Errorf("unable to get any supported resources from server. Error: %v.", err)
 		return nil, err
 	}
```
Comment (RainbowMango) [inline pkg/controllers/status/cluster_status_controller.go:388]: Can you split a PR for the partially APIs? This is our key mission, and shouldn't be blocked by the optional new condition.

File: pkg/util/membercluster_client.go (hunk 5 above)
Comment (duanmengkk) [reply pkg/util/membercluster_client.go:39]: It is a `aggregatorclient` for create a apiservice resource,which is needed for e2e test,not for list api-resources.

File: pkg/controllers/status/cluster_status_controller.go (hunk 6 above)
Comment (duanmengkk) [reply pkg/controllers/status/cluster_status_controller.go:388]: > Can you split a PR for the partially APIs?

I am not clearly  about what you sayed.Do you mean this pr should not contain e2e test,just focus on the feature of  `metadataSufficient `status.Or is there a problem in line 385 and 386 ?

Comment (duanmengkk) [reply pkg/controllers/status/cluster_status_controller.go:388]: > Do you mean this pr should not contain e2e test,just focus on the feature of `metadataSufficient `status

If my understand is right,I will split this PR in tree:1. Add metadataSufficient  status 2. Add output of metadataSufficient  in `kubectl get cluster` 3.Add new e2e test for metadataSufficient  status

Comment (duanmengkk) [reply pkg/controllers/status/cluster_status_controller.go:388]: > Can you split a PR for the partially APIs? This is our key mission, and shouldn't be blocked by the optional new condition.

If you do not mind,Can you describe it in Chinese?

File: pkg/util/membercluster_client.go (hunk 5 above)
Comment (XiShanYongYe-Chang) [reply pkg/util/membercluster_client.go:39]: This client appears to be used to simulate creating and getting APIService resources in a member cluster.

File: pkg/controllers/status/cluster_status_controller.go (hunk 6 above)
Comment (RainbowMango) [reply pkg/controllers/status/cluster_status_controller.go:388]: I mean the change for `func getAPIEnablements` is what we urgently needed. We can do it in a separate PR.

For the `MetadataSufficient` condition, I'm a little bit hesitant as it seems to lack enough benefits to users.

Comment (duanmengkk) [reply pkg/controllers/status/cluster_status_controller.go:388]: Thanks. I got it

Comment (duanmengkk) [reply pkg/controllers/status/cluster_status_controller.go:388]: A separate PR in https://github.com/karmada-io/karmada/pull/1968 
/cc @RainbowMango

**Pull request closed without merging**